File: 3k-1s/architecture/laba4/benchit/kernel/numerical/jacobi/Java/0/0/sequential/Jacobi.java
/********************************************************************
 * BenchIT - Performance Measurement for Scientific Applications
 * Contact: developer@benchit.org
 *
 * $Id: Jacobi.java 1 2009-09-11 12:26:19Z william $
 * $URL: svn+ssh://william@rupert.zih.tu-dresden.de/svn-base/benchit-root/BenchITv6/kernel/numerical/jacobi/Java/0/0/sequential/Jacobi.java $
 * For license details see COPYING in the package base directory
 *******************************************************************/
/* Kernel: Jacobi algorithm measuring FLOPS (ij, ji) for change of
 *         dimension
 *******************************************************************/

/* This class definies the measured algorithm. */
public class Jacobi {
	/* constants */
	protected DataObject dataObject;

	public Jacobi( DataObject dataObject ) {
		this.dataObject = dataObject;
	}

	/* This methode inititializes static attributes that need
	 * to be initialized only once. */
	public void twodinit() {
		dataObject.mitsdone = 0;
		dataObject.diffnorm = 0.0;
		/* boundary condition: a and b have value 1 on lower bound (0 < x < 1, y = 0) */
		for ( int i = 0; i < dataObject.maxn; i++ ) {
			for ( int j = 0; j < dataObject.maxn; j++ ) {
				dataObject.a[i * dataObject.maxn + j] = 0.0;
				dataObject.b[i * dataObject.maxn + j] = 0.0;
				dataObject.f[i * dataObject.maxn + j] = 0.0;
			}
		}
		for ( int j = 1; j < dataObject.maxn - 1; j++ ) {
			int index = j * dataObject.maxn;
			dataObject.a[index] = 1.0;
			dataObject.b[index] = 1.0;
		}
		for ( int i = 1; i < dataObject.maxn - 1; i++ ) {
			int index = i;
			dataObject.a[index] = 1.0;
			dataObject.b[index] = 1.0;
		}
	}

	/* Computation of the stencil writing to b[]. */
	public void sweep2d_afnb_ji() {
		int index, i, j;
		for ( j = 1; j < dataObject.maxn - 1; j++ ) {
			for ( i = 1; i < dataObject.maxn - 1; i++ ) {
				index = j * dataObject.maxn + i;
				dataObject.b[index] = 0.25 *
					( dataObject.a[index - dataObject.maxn]
               + dataObject.a[index - 1] + dataObject.a[index + 1]
               + dataObject.a[index + dataObject.maxn] ) -
					dataObject.h * dataObject.h * dataObject.f[index];
			}
		}
	}

	public void sweep2d_afnb_ij() {
		int index, i, j;
		for ( i = 1; i < dataObject.maxn - 1; i++ ) {
			for ( j = 1; j < dataObject.maxn - 1; j++ ) {
				index = j * dataObject.maxn + i;
				dataObject.b[index] = 0.25 *
					( dataObject.a[index - dataObject.maxn]
               + dataObject.a[index - 1] + dataObject.a[index + 1]
               + dataObject.a[index + dataObject.maxn] ) -
					dataObject.h * dataObject.h * dataObject.f[index];
			}
		}
	}

	/** Computation of the stencil writing to a[]. */
	public void sweep2d_bfna_ji() {
		int index, i, j;
		for ( j = 1; j < dataObject.maxn - 1; j++ ) {
			for ( i = 1; i < dataObject.maxn - 1; i++ ) {
				index = j * dataObject.maxn + i;
				dataObject.a[index] = 0.25 *
					( dataObject.b[index - dataObject.maxn]
               + dataObject.b[index - 1] + dataObject.b[index + 1]
               + dataObject.b[index + dataObject.maxn] ) -
					dataObject.h * dataObject.h * dataObject.f[index];
			}
		}
	}
	public void sweep2d_bfna_ij() {
		int index, i, j;
		for ( i = 1; i < dataObject.maxn - 1; i++ ) {
			for ( j = 1; j < dataObject.maxn - 1; j++ ) {
				index = j * dataObject.maxn + i;
				dataObject.a[index] = 0.25 *
					( dataObject.b[index - dataObject.maxn]
               + dataObject.b[index - 1] + dataObject.b[index + 1]
               + dataObject.b[index + dataObject.maxn] ) -
					dataObject.h * dataObject.h * dataObject.f[index];
			}
		}
	}

	/* Calculation of residual error between sub-arrays a[] and b[]. */
	public double diff2d_ji() {
		int index, i, j;
		double diff = 0.0;
		double sum = 0.0;
		for ( j = 1; j < dataObject.maxn - 1; j++ ) {
			for ( i = 1; i < dataObject.maxn - 1; i++ ) {
				index = j * dataObject.maxn + i;
				diff = dataObject.a[index] - dataObject.b[index];
				sum += diff * diff;
			}
		}
		return sum;
	}

	public double diff2d_ij() {
		int index, i, j;
		double diff = 0.0;
		double sum = 0.0;
		for ( i = 1; i < dataObject.maxn - 1; i++ ) {
			for ( j = 1; j < dataObject.maxn - 1; j++ ) {
				index = j * dataObject.maxn + i;
				diff = dataObject.a[index] - dataObject.b[index];
				sum += diff * diff;
			}
		}
		return sum;
	}

	/* Helps the programmer to write less code ;-). */
	private void println( String msg ) {
		System.out.println( msg );
	}

	/* If you call this, it will print the b[]-array at any place
	 * during execution. */
	public void printStatB() {
	   StringBuffer sbb = new StringBuffer();
	   int j, i, index = 0;
	   sbb.append( "double[] b:\n[" );
	   for ( j = 0; j < dataObject.maxn; j++ ) {
		   for ( i = 0; i < dataObject.maxn; i++ ) {
			   index = j * dataObject.maxn + i;
			   String st = new String(
               ( new Double( dataObject.b[index] ) ).toString() );
			   sbb.append( "[" );
			   for ( int space = 1; space <= ( 10 - st.length() ); space++ ) {
			      sbb.append( " " );
            }
			   sbb.append( st + "]" );
			   if ( i < ( dataObject.maxn - 1 ) ) {
		   	   sbb.append( "," );
            }
			}
			if ( j < ( dataObject.maxn - 1 ) ) {
			   sbb.append("\n ");
         }
	    }
	    sbb.append( "]" );
	    println( sbb.toString() );
	}

	public void printStatA() {
	   StringBuffer sbb = new StringBuffer();
	   int j, i, index = 0;
	   sbb.append( "double[] a:\n[" );
	   for ( j = 0; j < dataObject.maxn; j++ ) {
			for ( i = 0; i < dataObject.maxn; i++ ) {
			   index = j * dataObject.maxn + i;
			   String st = new String(
               ( new Double( dataObject.a[index] ) ).toString() );
			   sbb.append( "[" );
			   for ( int space = 1; space <= ( 10 - st.length() ); space++ ) {
				   sbb.append( " " );
            }
			   sbb.append( st + "]" );
			   if ( i < ( dataObject.maxn - 1 ) ) {
		      	sbb.append(",");
            }
			}
			if ( j < ( dataObject.maxn - 1 ) ) {
            sbb.append( "\n " );
         }
	    }
	    sbb.append( "]" );
	    println( sbb.toString() );
	}

	public void run_ji() {
      /* here is defined the job to do */
		for ( int it = 0; it < dataObject.maxIterations; it++ ) {
         /* main loop = iteration */
			sweep2d_bfna_ji();
			sweep2d_afnb_ji();
			dataObject.diffnorm = diff2d_ji();
			dataObject.mitsdone++;
			if ( dataObject.diffnorm < 1.3e-3 ) break;
			dataObject.diffnorm = 0.0;
		}
		if ( dataObject.diffnorm >= 1.3e-3 ) {
//			println( "failed to converge" );
		}
	}

	public void run_ij() {
      /* here is defined the job to do */
		for ( int it = 0; it < dataObject.maxIterations; it++ ) {
         /* main loop = iteration */
			sweep2d_bfna_ij();
			sweep2d_afnb_ij();
			dataObject.diffnorm = diff2d_ij();
			dataObject.mitsdone++;
			if ( dataObject.diffnorm < 1.3e-3 ) break;
			dataObject.diffnorm = 0.0;
		}
		if ( dataObject.diffnorm >= 1.3e-3 ) {
//			println( "failed to converge" );
		}
	}
}

File: 3k-1s/architecture/laba4/benchit/kernel/numerical/jacobi/C/0/0/1_malloc/jacobi.c
/********************************************************************
 * BenchIT - Performance Measurement for Scientific Applications
 * Contact: developer@benchit.org
 *
 * $Id: jacobi.c 1 2009-09-11 12:26:19Z william $
 * $URL: svn+ssh://william@rupert.zih.tu-dresden.de/svn-base/benchit-root/BenchITv6/kernel/numerical/jacobi/C/0/0/1_malloc/jacobi.c $
 * For license details see COPYING in the package base directory
 *******************************************************************/
/* Kernel: Jacobi algorithm measuring FLOPS (ij, ji) for change of
 *         dimension, one malloc for biggest dimension
 *******************************************************************/

#include "jacobi.h"

void printstat(double *, myinttype);
void sweep2da_ji(mydata_t *);
void sweep2da_ij(mydata_t *);
void sweep2db_ji(mydata_t *);
void sweep2db_ij(mydata_t *);
double diff2d_ji(mydata_t *);
double diff2d_ij(mydata_t *);

/* use this to print out one of the matrices */
void printstat(double *v, myinttype size) {
   myinttype i, j, index;
   for (j = 0; j < size; j++) {
      for (i = 0; i < size; i++) {
         index = j * size + i;
         printf("[%d:%d]=%1.3f ", i, j, v[index]);
      }
      printf("\n");
   }
   printf("\n");
}

/* specifies the work to do */
void jacobi_routine_ji(mydata_t * mdp) {
   myinttype it = 0;
   for (it = 0; it < mdp->mits; it++) {
      mdp->diffnorm = 0.0;
      sweep2db_ji(mdp);
      sweep2da_ji(mdp);
      mdp->diffnorm = diff2d_ji(mdp);
      if (mdp->diffnorm < 1.3e-3) {
         mdp->mitsdone = it;
         break;
      }
   }
   if (mdp->diffnorm >= 1.3e-3) {
//      printf("failed to converge!\n\n");
      mdp->mitsdone = it;
   }
}

/* specifies the work to do */
void jacobi_routine_ij(mydata_t * mdp) {
   myinttype it = 0;
   for (it = 0; it < mdp->mits; it++) {
      mdp->diffnorm = 0.0;
      sweep2db_ij(mdp);
      sweep2da_ij(mdp);
      mdp->diffnorm = diff2d_ij(mdp);
      if (mdp->diffnorm < 1.3e-3) {
         mdp->mitsdone = it;
         break;
      }
   }
   if (mdp->diffnorm >= 1.3e-3) {
//      printf("failed to converge!\n\n");
      mdp->mitsdone = it;
   }
}

void twodinit(mydata_t * mdp) {
   myinttype i = 0, j = 0, index = 0;
   mdp->diffnorm = 0.0;
   mdp->mitsdone = 0;
   for (j = 0; j < mdp->maxn; j++) {
      for (i = 0; i < mdp->maxn; i++) {
         index = j * mdp->maxn + i;
         mdp->a[index] = 0.0;
         mdp->b[index] = 0.0;
         mdp->f[index] = 0.0;
      }
   };
   for (i = 1; i < (mdp->maxn - 1); i++) {
      index = i;
      mdp->a[index] = 1.0;
      mdp->b[index] = 1.0;
   }
   for (j = 1; j < (mdp->maxn - 1); j++) {
      index = mdp->maxn * j;
      mdp->a[index] = 1.0;
      mdp->b[index] = 1.0;
   }
}

void sweep2da_ji(mydata_t * mdp) {
   myinttype index = 0, i = 0, j = 0;
   for (j = 1; j < mdp->maxn - 1; j++) {
      for (i = 1; i < mdp->maxn - 1; i++) {
         index = j * mdp->maxn + i;
         mdp->b[index] =
            0.25 * (mdp->a[index - mdp->maxn] + mdp->a[index - 1] +
                    mdp->a[index + mdp->maxn] + mdp->a[index + 1]) -
            mdp->h * mdp->h * mdp->f[index];
      }
   }
}
void sweep2da_ij(mydata_t * mdp) {
   myinttype index = 0, i = 0, j = 0;
   for (i = 1; i < mdp->maxn - 1; i++) {
      for (j = 1; j < mdp->maxn - 1; j++) {
         index = j * mdp->maxn + i;
         mdp->b[index] =
            0.25 * (mdp->a[index - mdp->maxn] + mdp->a[index - 1] +
                    mdp->a[index + mdp->maxn] + mdp->a[index + 1]) -
            mdp->h * mdp->h * mdp->f[index];
      }
   }
}

void sweep2db_ji(mydata_t * mdp) {
   myinttype index = 0, i = 0, j = 0;
   for (j = 1; j < mdp->maxn - 1; j++) {
      for (i = 1; i < mdp->maxn - 1; i++) {
         index = j * mdp->maxn + i;
         mdp->a[index] =
            0.25 * (mdp->b[index - mdp->maxn] + mdp->b[index - 1] +
                    mdp->b[index + mdp->maxn] + mdp->b[index + 1]) -
            mdp->h * mdp->h * mdp->f[index];
      }
   }
}
void sweep2db_ij(mydata_t * mdp) {
   myinttype index = 0, i = 0, j = 0;
   for (i = 1; i < mdp->maxn - 1; i++) {
      for (j = 1; j < mdp->maxn - 1; j++) {
         index = j * mdp->maxn + i;
         mdp->a[index] =
            0.25 * (mdp->b[index - mdp->maxn] + mdp->b[index - 1] +
                    mdp->b[index + mdp->maxn] + mdp->b[index + 1]) -
            mdp->h * mdp->h * mdp->f[index];
      }
   }
}

double diff2d_ji(mydata_t * mdp) {
   myinttype index = 0, i = 0, j = 0;
   double diff = 0.0, sum = 0.0;
   for (j = 1; j < mdp->maxn - 1; j++) {
      for (i = 1; i < mdp->maxn - 1; i++) {
         index = j * mdp->maxn + i;
         diff = mdp->a[index] - mdp->b[index];
         sum += diff * diff;
      }
   }
   return sum;
}
double diff2d_ij(mydata_t * mdp) {
   myinttype index = 0, i = 0, j = 0;
   double diff = 0.0, sum = 0.0;
   for (i = 1; i < mdp->maxn - 1; i++) {
      for (j = 1; j < mdp->maxn - 1; j++) {
         index = j * mdp->maxn + i;
         diff = mdp->a[index] - mdp->b[index];
         sum += diff * diff;
      }
   }
   return sum;
}

